### PR TITLE
Make Fitting Parameter More Pydantic

### PR DESCRIPTION
**Pull request by eloiseyang**

### CODE REVIEW

File: lcls_tools/common/data_analysis/fit/method_base.py [hunk 1]
```diff
@@ -17,118 +66,70 @@ class MethodBase(ABC):
...
     """
 
-    param_names: list = None
-    param_bounds: np.ndarray = None
-
-    def __init__(self):
-
-        self.init_values: dict = None
-        self.fitted_params_dict: dict = None
+    model_parameters: ModelParameters = None
```
Comment (eloiseyang) [inline lcls_tools/common/data_analysis/fit/method_base.py:69]: Could you rename this to `parameters`? As it stands we end up writing `model.model_parameters`, and the repeat word is a bit awkward.

File: lcls_tools/common/data_analysis/fit/methods.py [hunk 2]
```diff
@@ -1,101 +1,102 @@
...
-    param_names: list = ["mean", "sigma", "amplitude", "offset"]
-    param_bounds: np.ndarray = np.array(
-        [
-            [0.01, 1.0],
-            [0.01, 5.0],
-            [0.01, 1.0],
-            [0.01, 1.0],
-        ]
-    )
+    model_parameters = gaussian_parameters
```
Comment (eloiseyang) [inline lcls_tools/common/data_analysis/fit/methods.py:31]: Could you rename this variable to `parameters`?

**Review by eloiseyang — COMMENTED** [review]
I like the Parameter and ModelParameters classes. It looks like they make defining parameters easier. I made comments for some syntax suggestions, as well as some places where we can use named keys instead of numbers. Once we get those resolved we can move it into the main PR.

Thanks!!!

File: lcls_tools/common/data_analysis/fit/method_base.py [hunk 3]
```diff
@@ -1,7 +1,56 @@
...
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    name: str
+    parameters: dict[str, Parameter]
+
+    @property
+    def bounds(self):
+        return np.vstack(
+            [
+                np.array(self.parameters[parameter].bounds)
+                for parameter in self.parameters
```
[the remark this answers is not in the document]
Comment (phys-cgarnier) [reply lcls_tools/common/data_analysis/fit/method_base.py:25]: Yes, will do.

File: lcls_tools/common/data_analysis/fit/method_base.py [hunk 4]
```diff
@@ -17,118 +66,70 @@ class MethodBase(ABC):
...
+    def forward(
+        self, x: np.ndarray, method_parameter_dict: dict[str, float]
+    ) -> np.ndarray:
+        method_parameter_list = np.array(
+            [
+                method_parameter_dict[parameter_name]
+                for parameter_name in self.model_parameters.parameters
+            ]
+        )
+        return self._forward(x, method_parameter_list)
```
[the remark this answers is not in the document]
Comment (phys-cgarnier) [reply lcls_tools/common/data_analysis/fit/method_base.py:86]: Unfortunately we need the function _forward. It is essentially what is being called iteratively in the fit method by scipy.optimize.minimze. The "fun" argument of this method needs to be a callable function that takes a list of values. In this case that list is our ordered gaussian parameters.

File: lcls_tools/common/data_analysis/fit/methods.py [hunk 5]
```diff
@@ -1,101 +1,102 @@
...
-            self.param_names[2]: amplitude_prior,
-            self.param_names[3]: offset_prior,
-        }
-        return self.priors
+        offset_prior = norm(init_values["offset"], 0.5)
+        parameters = [parameter for parameter in init_values]
+        priors = dict(
+            zip(parameters, [mean_prior, sigma_prior,
+                             amplitude_prior, offset_prior])
+        )
```
[the remark this answers is not in the document]
Comment (phys-cgarnier) [reply lcls_tools/common/data_analysis/fit/methods.py:82]: Yeah we can do that.

File: lcls_tools/common/data_analysis/fit/method_base.py (hunk 1 above)
Comment (phys-cgarnier) [reply lcls_tools/common/data_analysis/fit/method_base.py:69]: Sure, will do.

File: lcls_tools/common/data_analysis/fit/method_base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eloiseyang) [reply lcls_tools/common/data_analysis/fit/method_base.py:86]: Okay, that makes sense. I thought that might be the case here. Thanks for the clarification.

**Pull request merged**